### PR TITLE
Reverted changes and added function collate_pad

### PR DESCRIPTION
**Pull request by eherrewijnen**
Added function collate_pad to collate.py, for usage with varying-length inputs.
Example usage: 
```
from torch.utils.data import DataLoader
from torch.utils/data._utils.collate import collate_pad
dataloader = DataLoader(dataset, collate_fn=collate_pad)
```

### CODE REVIEW

File: torch/utils/data/_utils/collate.py [hunk 1]
```diff
@@ -201,6 +201,67 @@ def collate_str_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Typ
...
+    Args:
+        batch: a single batch to be padded and collated
+        batch_first (bool, optional): output will be in ``B x T x *`` if True, or in
+            ``T x B x *`` otherwise. Default: False.
+        padding_value (float, optional): value for padded elements. Default: 0.
+    Examples:
+        >>> # xdoctest: +SKIP
+        >>> # Example with a batch of `int`s:
+        >>> collate_pad([[1,2,3], [4], [5,6,7,8]])
+        tensor([[1,2,3,0], [4], [5,6,7,8]])
```
Comment (ejguan) [inline torch/utils/data/_utils/collate.py:226]: Should the result be `tensor([[1,2,3,0], [4,0,0,0], [5,6,7,8]])`?

File: torch/utils/data/_utils/collate.py [hunk 2]
```diff
@@ -201,6 +201,67 @@ def collate_str_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Typ
...
+    """
+    first_elem = batch[0]
+    elem_type = type(first_elem)
+    if isinstance(first_elem, np.ndarray):
+        batch = [torch.from_numpy(b) for b in batch]
+
+    if any(first_elem.shape != elem_.shape for elem_ in batch):
```
Comment (ejguan) [inline torch/utils/data/_utils/collate.py:233]: This is not working, IMO. You are assuming each element would be `Tensor` with `shape` attribute in `batch`.

It's even not working when the element is `mapping`.

**Review by ejguan — CHANGES_REQUESTED** [review]

File: torch/utils/data/_utils/collate.py (hunk 2 above)
Comment (eherrewijnen) [reply torch/utils/data/_utils/collate.py]: Good point! I've updated the order now.

File: torch/utils/data/_utils/collate.py [hunk 3]
```diff
@@ -200,6 +200,69 @@ def collate_str_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Typ
...
+        batch = [torch.from_numpy(b) for b in batch]
+        return torch.nn.utils.rnn.pad_sequence(
+            batch, batch_first=batch_first, padding_value=padding_value
+        )
+    elif isinstance(first_elem, torch.Tensor) and any(first_elem.shape != elem_.shape for elem_ in batch):
+        return torch.nn.utils.rnn.pad_sequence(
+            batch, batch_first=batch_first, padding_value=padding_value
+        )
+    else:
+        return default_collate(batch)
```
Comment (ejguan) [inline torch/utils/data/_utils/collate.py:262]: Can we raise `Exception` here with message like type not supported and please implement custom padding collate? IMHO, falling back to `default_collate` makes users confused by the expected behavior.

File: torch/utils/data/_utils/collate.py [hunk 4]
```diff
@@ -200,6 +200,69 @@ def collate_str_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Typ
...
+
```
Comment (ejguan) [inline torch/utils/data/_utils/collate.py:263]: nit: two empty lines between functions.
